### PR TITLE
[MCC-170850] Add ExpendoStateService

### PR DESCRIPTION
**Pull request by junshao**
- Added `ExpendoStateService` to keep states of static class or instance. 
- Nuget package 1.0.0 was pushed onto the medidata gallery. http://nuget.imedidata.net/feed/mdsol/package/nuget/Medidata.Cloud.Thermometer.RaveCommon

@art2003 @echen-mdsol @vcapers-mdsol @chenghuang-mdsol Please review and merge if OK. Thanks.

### `ExpendoStateService` for static classes
```cs
// Set state
var stateService = DIContainer.Resolve<IExpendoStateService>();
stateService.ForClass<StaticClass>().Set("State", "Running");

// Get state
var value = stateService.ForClass<StaticClass>().Get("State");
Assert.AreEqual("Running", value);
```
Or you can use equivalent non-generic way as below.
```cs
// Set state
var stateService = DIContainer.Resolve<IExpendoStateService>();
stateService.ForClass(typeof(StaticClass)).Set("State", "Running");

// Get state
var value = stateService.ForClass(typeof(StaticClass)).Get("State");
Assert.AreEqual("Running", value);
```

### `ExpendoStateService` for instances
```cs
var cat, mouse;
var stateService = DIContainer.Resolve<IExpendoStateService>();

stateService.ForInstance(cat).Set("Name", "Tom");
stateService.ForInstance(mouse).Set("Name", "Jerry");

var catName = stateService.ForInstance(cat).Get("Name");
var mouseName = stateService.ForInstance(mouse).Get("Name");

Assert.AreEqual("Tom", catName);
Assert.AreEqual("Jerry", mouseName);
```

@chenghuang-mdsol You can refer to the README.md file and feature files for the detail usages.

## Unit tests
image

## SpecFlow tests
image

### CODE REVIEW

File: Mct.RaveCommon/ExpendoState/ExpendoStateConcurrentStorage.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+namespace Medidata.Cloud.Thermometer.RaveCommon.ExpendoState
+{
+    internal class ExpendoStateConcurrentStorage : IExpendoStateStorage
+    {
+        internal readonly ConcurrentDictionary<int, Dictionary<string, object>> PropDic =
+            new ConcurrentDictionary<int, Dictionary<string, object>>();
+
+        public IDictionary<string, object> GetStorage(int targetIdentity)
+        {
+            return PropDic.AddOrUpdate(targetIdentity, new Dictionary<string, object>(), (k, v) => v);
```
Comment (echen-mdsol) [inline Mct.RaveCommon/ExpendoState/ExpendoStateConcurrentStorage.cs]: I think this should be GetOrAdd instead of AddOrUpdate?

Comment (junshao) [inline Mct.RaveCommon/ExpendoState/ExpendoStateConcurrentStorage.cs]: You are right. Thanks. Will change.

File: Mct.RaveCommon/ExpendoStateService.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        public virtual IExpendoStateAccessor ForClass<T>() where T : class
+        {
+            return ForClass(typeof (T));
+        }
+
+        public IExpendoStateAccessor ForClass(Type type)
+        {
+            if (type == null) throw new ArgumentNullException("type");
```
Comment (art2003) [inline Mct.RaveCommon/ExpendoStateService.cs]: Do we need to check here that type IsClass?

Comment (junshao) [inline Mct.RaveCommon/ExpendoStateService.cs]: Good. Will add. Thanks

Comment (junshao) [inline Mct.RaveCommon/ExpendoStateService.cs]: Comment reflected.

**Comment by art2003** [comment]
LGTM

**Pull request merged**